### PR TITLE
Course Creation and Specification

### PR DESCRIPTION
**Pull request by spencerj411**
Completed Issue #27 (Course Creation and Specification (5)).

Can access this page at `http://127.0.0.1:8000/manage_courses/`. It displays a form in 3 sections (for this user story, only 'Add Course' has been implemented).

**The acceptance criterion:** 
> If the requisite/incompatible courses aren’t in the system after the user saves the new course, the user is prompted to fill in the course details for the requisite/incompatible course

has been satisfied by listing all of the courses in the database instead of giving the option to the user to type the courses manually.
> Notify the user if they are attempting to add an existing course

was implemented with a basic HttpResponse.

**Small things:**
1. I was not sure how to concatenate 3 strings together hence `{{ course.code }}_{{ course.year }}` appeared 4 times in `managecourses.html` (solved)
2. Creation of courses can be tested by checking `http://127.0.0.1:8000/list/?view=Course`
3. Screenshot of UI  (outdated, check below for a more updated screenshot)
image

### CODE REVIEW

**Review by JackSekoranja — CHANGES_REQUESTED** [review]
Form looks really good, but since adding and editing should be the same form we may be able to simplify it using a similar trick to my `list` page, so I think leave the additional forms out of it for now. There were also a few minor changes that I will suggest below.

File: cassdegrees/ui/views.py
Comment (JackSekoranja) [inline cassdegrees/ui/views.py]:
```suggestion
        model_api_url = request.build_absolute_uri('/api/model/course/')
```

This Makes it a relative path, meaning we won't have to change it later (We should probably do this in `sampleform` above too)

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [inline cassdegrees/templates/managecourses.html]:
```suggestion
                <input class="text tfull" id="number" type="number" name="year" min="1990" aria-required="true" required>
```

Changes will make the form stretch across to the end of the page. Also remove max year, it isn't needed.

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [inline cassdegrees/templates/managecourses.html]:
```suggestion
                <input class="text tfull" id="text" type="text" name="name" aria-required="true" required>
```

Makes the text box stretch to the end of the page, giving the user more space to type

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [inline cassdegrees/templates/managecourses.html]:
```suggestion
                <input class="text tfull" id="text" type="text" name="code" minlength="8" maxlength="8" size="8" aria-required="true" required>
```

Same as above (for consistency)

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [inline cassdegrees/templates/managecourses.html]: I can't work out how to make this stretch to the end of the screen, but it would be good to work out

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [inline cassdegrees/templates/managecourses.html]:
```suggestion
                <input class="text tfull" id="number" type="number" name="units" min="0" step="6" aria-required="true" required>
```

Courses can be worth 0 units, so this should be reflected here.
Stretch to end of form and removed max units for a course (Realistically it should only be 12-24 max but why restrict them?)

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [inline cassdegrees/templates/managecourses.html]:
```suggestion
                        <option value="{{ course.code }}">{{ course.code }}</option>
```

Course year is not relevant to requisites, so I have removed them (Need to ensure this is also done when passing the API to prevent repeat values). This also raises another important issue, which is that the requisites is not currently represented in the database. This may be a problem for @aidengeorge97 to solve

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [inline cassdegrees/templates/managecourses.html]:
```suggestion
                        <option value="{{ course.code }}">{{ course.code }}</option>
```

As above

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [inline cassdegrees/templates/managecourses.html]: Might be a good idea to remove everything after this line until the `{% endblock %}` line. `Delete` could have it's own form, but the `edit` form would be the exact same as the one above, only with pre-filled values and with the `value=patch` item. Ask me if unsure about how to approach for future stories

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [inline cassdegrees/templates/managecourses.html]: General note about comments in this page. When you comment with `<!-- ... -->` it appears when you view the page source, but when you use Django Template Comments (`{# ... #}`) they get deleted when the page is hosted.

File: cassdegrees/ui/views.py
Comment (JackSekoranja) [inline cassdegrees/ui/views.py:108]:
```suggestion
    courses = requests.get(request.build_absolute_uri('/api/model/course/?format=json')).json()
    courses = [{'code': course} for course in  set([x['code'] for x in courses])]
```

Convert courses to a unique set of only course codes. Please note that this is NOT a good solution and is intended to be a single line fix. Best way might be to try do a SQL Query for unique course codes, consult @Daniel-Jang for this

**Review by JackSekoranja — CHANGES_REQUESTED** [review]
These are my suggested changes

File: cassdegrees/templates/managecourses.html
Comment (spencerj411) [reply cassdegrees/templates/managecourses.html]: my thought process was so it leaves the user less of a chance to create a mistake when creating a course, hence adding max for both year and units.

File: cassdegrees/templates/managecourses.html
Comment (spencerj411) [reply cassdegrees/templates/managecourses.html]: my thought process was so it leaves the user less of a chance to create a mistake when creating a course, hence adding max for both year. My original intention was to make it dynamic but will do that at a later date once the main work has been approved and if everyone is ok with this.

File: cassdegrees/templates/managecourses.html
Comment (spencerj411) [reply cassdegrees/templates/managecourses.html]: good idea

File: cassdegrees/templates/managecourses.html
Comment (spencerj411) [reply cassdegrees/templates/managecourses.html]: not particularly fond of having all the textboxes stretch to the other side of the page, especially when some input don't need to be that long. Having said that, after visiting [http://webpublishing.anu.edu.au/web-style-guide/forms.php](url) it seems like ANU likes it so I'm ok with this.

File: cassdegrees/templates/managecourses.html
Comment (spencerj411) [reply cassdegrees/templates/managecourses.html]: will do

File: cassdegrees/ui/views.py
Comment (spencerj411) [reply cassdegrees/ui/views.py:108]: will do

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [reply cassdegrees/templates/managecourses.html]: So it looks like the ANU `tfull` CSS just makes the width of text boxes 70%. Maybe a solution is to create a custom CSS field doing the same thing for select boxes

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [reply cassdegrees/templates/managecourses.html]: Alternatively maybe you could reposition year and units to be at the bottom and leave those smaller, while making the other text fields/select fields larger. Might need further discussion with team

File: cassdegrees/templates/managecourses.html
Comment (JackSekoranja) [reply cassdegrees/templates/managecourses.html]: I still don't think we should restrict the software like that though. Maybe we could ask the user for confirmation if the entered year is too distant from the current year (Implementation may be too complicated for the resulting gain though)

**Comment by spencerj411** [comment]
UI Screenshot after Jack's suggestions
image

**Comment by spencerj411** [comment]
UI after implementing Jack's suggestions
image

**Review by JackSekoranja — APPROVED** [review]
Changes look good. There is an issue where the CSS for `<select>` objects look misaligned but not sure if we can do much about it.

File: cassdegrees/static/css/style.css
Comment (j-selby) [inline cassdegrees/static/css/style.css:36]: Not specific to this PR, but:

http://webpublishing.anu.edu.au/web-style-guide/style-sampler.php
`Please avoid using inline styles or creating custom style sheets`

We should be justifying each use of rules in this file.

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py:108]: Wait, is this request made even for POST requests, which don't appear to use 'courses' at all? This doesn't seem right.

**Review by j-selby — CHANGES_REQUESTED** [review]

File: cassdegrees/ui/views.py
Comment (spencerj411) [reply cassdegrees/ui/views.py:108]: @JackSekoranja

File: cassdegrees/static/css/style.css
Comment (spencerj411) [reply cassdegrees/static/css/style.css:36]: Not sure if it was mentioned above but it didn't seem like there was a `<select> `version of `tfull` in their css file hence the creation of new rules.

**Comment by spencerj411** [comment]
also added a proper implementation of the error/success message when submitting a form (based off #59) using ANU's style guide.

image

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py]: Fix up spacing here?

File: cassdegrees/ui/views.py
Comment (j-selby) [inline cassdegrees/ui/views.py]: We shouldn't be using this for these forms - we only need POST and GET. Get rid of this if statement entirely IMO.

File: cassdegrees/ui/views.py
Comment (j-selby) [reply cassdegrees/ui/views.py:108]: No longer relevant - main template is now always rendered.

File: cassdegrees/ui/views.py
Comment (Daniel-Jang) [reply cassdegrees/ui/views.py]: actual_request is now deleted.

**Review by Daniel-Jang — CHANGES_REQUESTED** [review]
This form, as is, can't handle course prereqs properly as they are currently done (e.g. COMPXXXX and COMPXXXX or COMPXXXX).

Also, this needs to be tabbed with Edit, Add and Delete tabs with this spec.:

"When you click on manage courses, there are 3 tabs: Edit, Add and Delete. Screen defaults to edit courses tab, which would have the table of courses with a link that brings out a form that you can change and click submit.
Add tab would just be the Spencer's page in it's current form (look at most recent pic)
Delete would be like the edit page but instead the link submits a delete request to the API."

I have already had conversations with Jack and Spencer and they believe it is a good idea. Spencer has already started implementing this and I will be trying to help him so I can work on my user story: https://github.com/cass-degrees/CASS-Degrees-Code/issues/31

EDIT: Had a chat with Jack and we will have to have a discussion later with Aiden on how to store the requisites properly. Currently, there is no method of storing it properly. Hence, I will approve Spencer's ticket as there is nothing he can do in this moment.

**Review by Daniel-Jang — APPROVED** [review]

**Pull request merged**
